Help_pro refurbishment and code cleanup: part 1

maynardGK opened the pull request:
The help routine had been included with the basic_pro set of routines
and it grew to become the largest single routine with the most keywords.
It grew in increments and so is a challenge to decipher.
These changes put much of the formatting code in subroutines called from
help_pro, allowing the flow and logic of help_pro to be more approachable.
Some new features of help will  also be present:
help,/heap
COMMON, OBJECT, METHODS keywords.
Output=<variable> is implemented in many more places.
This is part 1, which is maybe 1/3 of the way finished to help_pro() which is
pretty much the same as it was.

gdlhelp
Greg Jung (aka maynardGK)
gvjung@gmail.com

maynardGK committed:
Help_pro refurbishment and code cleanup: part 1

The help routine had been included with the basic_pro set of routines
and it grew to become the largest single routine with the most keywords.
It grew in increments and so is a challenge to decipher.
These changes put much of the formatting code in subroutines called from
help_pro, allowing the flow and logic of help_pro to be more approachable.
Some new features of help will  also be present:
help,/heap
COMMON, OBJECT, METHODS keywords.
Output=<variable> is implemented in many more places.
This is part 1, which is maybe 1/3 of the way finished to help_pro() which is
pretty much the same as it was.

gdlhelp
Greg Jung (aka maynardGK)
gvjung@gmail.com

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
# Explicit paths specified without -i or -o; assuming --only paths...
# On branch master
# Your branch is up-to-date with 'origin/master'.
#
# Changes to be committed:
#	modified:   src/gdlhelp.cpp
#
# Changes not staged for commit:
#	modified:   CMakeLists.txt
#	modified:   src/devicewin.hpp
#
# Untracked files:
#	dotouch.sh
#	"h\033:"
#	src/gdl
#	testsuite/file_etc/
#	testsuite/launchtest.c
#	touchsrc.sh
#

slayoo commented:
This patch causes several tests to fail when gdl is compiled in Debug mode - see logs of Travis accessible from the "Details" link above

maynardGK commented:
Yes this is only part1 and not expected to pass test - just compiles.  Changes were so extensive
I wanted a record of intermediate state.  Working my way through it...

maynardGK committed:
Help_pro refurbishment and code cleanup: part 2 of 2 parts

This is part 2, all of the modifications made to the old version are incorporated, functionally.
Some new features of help will  also be present:
   Output=<variable> is implemented in many more places.
   names = <reg expression> used with other output to restrict the matches found.
   HEAP numerates object and pointer heap, also more detail if FULL is specified.
   COMMON will list all common blocks anywhere in program
   OBJECT, METHODS keywords for use with objects.
   DEBUG  will set or unset a debug flag that persists for session.

Other behavior changes:
 - For undefined variables, no output is generated unless /FULL keyword is also present.
 - the /STRUCT keyword, which is in effect if a structure is the sole parameter, will
  invoke a recursive description of any sub-structures.  The one-line description
  will always be present, before the detailed structure listing.

testsuite/test_help.pro is a very rudimentary start in order to provide
 an entry for Makefile.am

gdlhelp
Greg Jung (maynardGK)
gvjung@gmail.com

maynardGK commented:
Closing this PR and full set will be found in gdlhelp_fin branch.
